File: frontend/src/services/taskService.js
/**
 * Task Service - Serviço para gerenciamento de tarefas com Taskscore.
 * 
 * Sistema de gamificação inspirado na ADVBOX.
 */

import api from './api';

/**
 * Tipos de tarefa disponíveis
 */
export const TASK_TYPES = {
  audiencia: { label: 'Audiência', score: 100, icon: '⚖️' },
  prazo_fatal: { label: 'Prazo Fatal', score: 90, icon: '🚨' },
  peticao: { label: 'Petição', score: 80, icon: '📄' },
  analise: { label: 'Análise', score: 70, icon: '🔍' },
  diligencia: { label: 'Diligência', score: 60, icon: '🏃' },
  prazo_comum: { label: 'Prazo Comum', score: 50, icon: '📅' },
  reuniao: { label: 'Reunião', score: 40, icon: '👥' },
  contato_cliente: { label: 'Contato Cliente', score: 30, icon: '📞' },
  outro: { label: 'Outro', score: 25, icon: '📌' },
  administrativo: { label: 'Administrativo', score: 20, icon: '📋' },
};

/**
 * Prioridades e seus multiplicadores
 */
export const TASK_PRIORITIES = {
  baixa: { label: 'Baixa', multiplier: 0.8, color: 'gray' },
  media: { label: 'Média', multiplier: 1.0, color: 'blue' },
  alta: { label: 'Alta', multiplier: 1.3, color: 'orange' },
  urgente: { label: 'Urgente', multiplier: 1.5, color: 'red' },
};

/**
 * Status das tarefas
 */
export const TASK_STATUSES = {
  pendente: { label: 'Pendente', color: 'yellow' },
  em_andamento: { label: 'Em Andamento', color: 'blue' },
  concluida: { label: 'Concluída', color: 'green' },
  cancelada: { label: 'Cancelada', color: 'gray' },
};

/**
 * Níveis de alerta para prazos
 */
export const ALERT_LEVELS = {
  normal: { label: 'Normal', color: 'green', bgColor: 'bg-green-100', textColor: 'text-green-800' },
  attention: { label: '7 dias', color: 'blue', bgColor: 'bg-blue-100', textColor: 'text-blue-800' },
  warning: { label: '3 dias', color: 'yellow', bgColor: 'bg-yellow-100', textColor: 'text-yellow-800' },
  critical: { label: '1 dia', color: 'orange', bgColor: 'bg-orange-100', textColor: 'text-orange-800' },
  fatal: { label: 'Hoje', color: 'red', bgColor: 'bg-red-100', textColor: 'text-red-800' },
  overdue: { label: 'Vencida', color: 'red', bgColor: 'bg-red-200', textColor: 'text-red-900' },
};

/**
 * Calcula a pontuação de uma tarefa
 */
export function calculateTaskScore(taskType, priority) {
  const baseScore = TASK_TYPES[taskType]?.score || 25;
  const multiplier = TASK_PRIORITIES[priority]?.multiplier || 1.0;
  return Math.round(baseScore * multiplier);
}

/**
 * Service para API de tarefas
 */
const taskService = {
  /**
   * Lista tarefas com filtros
   */
  async list(filters = {}) {
    const params = new URLSearchParams();
    if (filters.status) params.append('status', filters.status);
    if (filters.case_id) params.append('case_id', filters.case_id);
    if (filters.client_id) params.append('client_id', filters.client_id);
    if (filters.assigned_to) params.append('assigned_to', filters.assigned_to);
    if (filters.alert_level) params.append('alert_level', filters.alert_level);
    if (filters.limit) params.append('limit', filters.limit);
    if (filters.offset) params.append('offset', filters.offset);
    
    const response = await api.get(`/tasks/?${params.toString()}`);
    return response.data;
  },

  /**
   * Busca uma tarefa por ID
   */
  async get(taskId) {
    const response = await api.get(`/tasks/${taskId}`);
    return response.data;
  },

  /**
   * Cria uma nova tarefa
   */
  async create(taskData) {
    const response = await api.post('/tasks/', taskData);
    return response.data;
  },

  /**
   * Atualiza uma tarefa
   */
  async update(taskId, taskData) {
    const response = await api.put(`/tasks/${taskId}`, taskData);
    return response.data;
  },

  /**
   * Remove uma tarefa
   */
  async delete(taskId) {
    const response = await api.delete(`/tasks/${taskId}`);
    return response.data;
  },

  /**
   * Marca uma tarefa como concluída
   */
  async complete(taskId) {
    const response = await api.post(`/tasks/${taskId}/complete`);
    return response.data;
  },

  /**
   * Lista tarefas vencidas
   */
  async getOverdue() {
    const response = await api.get('/tasks/overdue');
    return response.data;
  },

  /**
   * Lista próximas tarefas
   */
  async getUpcoming(days = 7) {
    const response = await api.get(`/tasks/upcoming?days=${days}`);
    return response.data;
  },

  /**
   * Busca estatísticas de tarefas
   */
  async getStats() {
    const response = await api.get('/tasks/stats');
    return response.data;
  },

  /**
   * Busca ranking do Taskscore
   */
  async getRanking(startDate = null, endDate = null) {
    const params = new URLSearchParams();
    if (startDate) params.append('start_date', startDate.toISOString());
    if (endDate) params.append('end_date', endDate.toISOString());
    
    const response = await api.get(`/tasks/taskscore/ranking?${params.toString()}`);
    return response.data;
  },

  /**
   * Busca o Taskscore do usuário atual
   */
  async getMyScore(startDate = null, endDate = null) {
    const params = new URLSearchParams();
    if (startDate) params.append('start_date', startDate.toISOString());
    if (endDate) params.append('end_date', endDate.toISOString());
    
    const response = await api.get(`/tasks/taskscore/my-score?${params.toString()}`);
    return response.data;
  },

  /**
   * Completa múltiplas tarefas
   */
  async batchComplete(taskIds) {
    const response = await api.post('/tasks/batch/complete', taskIds);
    return response.data;
  },

  /**
   * Atualiza status de múltiplas tarefas
   */
  async batchUpdateStatus(taskIds, newStatus) {
    const response = await api.post('/tasks/batch/update-status', {
      task_ids: taskIds,
      new_status: newStatus,
    });
    return response.data;
  },
};

export default taskService;

File: backend/app/services/task_service.py
"""
Service para gerenciamento de tarefas com gamificação (Taskscore).

Sistema inteligente com:
- Pontuação por tarefa concluída
- Níveis de alerta baseados em prazo
- Ranking de produtividade
"""

import logging
from typing import List, Optional, Dict
from datetime import datetime, timezone

from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.schemas import (
    TaskCreate, TaskUpdate, TaskStatus, TaskPriority, TaskType
)
from app.models.db_models import (
    Task as DBTask,
    TaskStatusEnum
)
from app.services.base_sql_service import BaseSQLService

logger = logging.getLogger(__name__)


class TaskService(BaseSQLService[DBTask, TaskCreate, TaskUpdate]):
    """
    Service para operações com tarefas jurídicas.
    
    Features:
    - CRUD de tarefas (herda de BaseSQLService)
    - Sistema de pontuação (Taskscore)
    - Alertas de prazos
    - Ranking de produtividade
    """
    
    model = DBTask
    
    # Pontuação por tipo de tarefa
    TASK_SCORES = {
        TaskType.AUDIENCIA: 100,
        TaskType.PETICAO: 80,
        TaskType.PRAZO_FATAL: 90,
        TaskType.PRAZO_COMUM: 50,
        TaskType.REUNIAO: 40,
        TaskType.DILIGENCIA: 60,
        TaskType.ANALISE: 70,
        TaskType.CONTATO_CLIENTE: 30,
        TaskType.ADMINISTRATIVO: 20,
        TaskType.OUTRO: 25,
    }
    
    # Multiplicadores por prioridade
    PRIORITY_MULTIPLIERS = {
        TaskPriority.BAIXA: 0.8,
        TaskPriority.MEDIA: 1.0,
        TaskPriority.ALTA: 1.3,
        TaskPriority.URGENTE: 1.5,
    }
    
    def _calculate_score(self, task_type: TaskType, priority: TaskPriority) -> int:
        """Calcula pontuação da tarefa baseado no tipo e prioridade."""
        base_score = self.TASK_SCORES.get(task_type, 25)
        multiplier = self.PRIORITY_MULTIPLIERS.get(priority, 1.0)
        return int(base_score * multiplier)
    
    def _get_alert_level(self, due_date: datetime) -> str:
        """
        Determina o nível de alerta baseado na proximidade do prazo.
        
        Returns:
            "overdue" - Vencido
            "fatal" - Vence hoje
            "critical" - Vence em 1 dia
            "warning" - Vence em 3 dias
            "attention" - Vence em 7 dias
            "normal" - Mais de 7 dias
        """
        now = datetime.now(timezone.utc)
        
        # Se due_date não tem timezone, assume UTC
        if due_date.tzinfo is None:
            due_date = due_date.replace(tzinfo=timezone.utc)
        
        diff = (due_date - now).days
        
        if diff < 0:
            return "overdue"
        elif diff == 0:
            return "fatal"
        elif diff == 1:
            return "critical"
        elif diff <= 3:
            return "warning"
        elif diff <= 7:
            return "attention"
        else:
            return "normal"
    
    async def create(
        self,
        db: AsyncSession,
        data: TaskCreate | dict,
        user_id: str,
        **extra_fields
    ) -> DBTask:
        """
        Cria uma nova tarefa com pontuação calculada.
        
        Args:
            db: Sessão do banco
            data: Dados da tarefa
            user_id: ID do usuário
            
        Returns:
            Tarefa criada
        """
        # Calcula pontuação
        score = self._calculate_score(data.task_type, data.priority)
        
        # Define nível de alerta inicial
        if data.due_date:
            alert_level = self._get_alert_level(data.due_date)
        else:
            alert_level = "normal"
        
        # Usa create do BaseSQLService com campos extras
        return await super().create(
            db=db,
            data=data,
            user_id=user_id,
            score=score,
            alert_level=alert_level
        )
    
    async def complete_task(
        self,
        db: AsyncSession,
        task_id: str,
        user_id: str,
        completed_by: Optional[str] = None
    ) -> Optional[DBTask]:
        """
        Marca uma tarefa como concluída e registra quem completou.
        
        Args:
            db: Sessão do banco
            task_id: ID da tarefa
            user_id: ID do usuário owner
            completed_by: ID do usuário que completou (pode ser diferente do owner)
        
        Returns:
            Tarefa atualizada ou None
        """
        try:
            task = await self.get(db, task_id, user_id)
            if not task:
                return None
            
            task.status = TaskStatusEnum.CONCLUIDA
            task.completed_at = datetime.now(timezone.utc)
            task.completed_by = completed_by or user_id
            
            await db.commit()
            await db.refresh(task)
            
            logger.info(f"Task {task_id} completed by {task.completed_by}")
            return task
            
        except Exception as e:
            logger.error(f"Error completing task {task_id}: {e}")
            await db.rollback()
            return None
    
    async def list_by_status(
        self,
        db: AsyncSession,
        user_id: str,
        status: TaskStatus,
        limit: int = 50,
        offset: int = 0
    ) -> List[DBTask]:
        """Lista tarefas por status."""
        return await self.list(
            db=db,
            user_id=user_id,
            limit=limit,
            offset=offset,
            filters={"status": TaskStatusEnum(status.value)},
            order_by="due_date"
        )
    
    async def list_by_case(
        self,
        db: AsyncSession,
        case_id: str,
        user_id: str,
        limit: int = 50
    ) -> List[DBTask]:
        """Lista tarefas de um caso específico."""
        return await self.list(
            db=db,
            user_id=user_id,
            limit=limit,
            filters={"case_id": case_id},
            order_by="due_date"
        )
    
    async def list_overdue(
        self,
        db: AsyncSession,
        user_id: str,
        limit: int = 50
    ) -> List[DBTask]:
        """Lista tarefas vencidas (overdue)."""
        try:
            now = datetime.now(timezone.utc)
            
            result = await db.execute(
                select(DBTask)
                .where(
                    DBTask.user_id == user_id,
                    DBTask.status != TaskStatusEnum.CONCLUIDA,
                    DBTask.status != TaskStatusEnum.CANCELADA,
                    DBTask.due_date < now
                )
                .order_by(DBTask.due_date)
                .limit(limit)
            )
            
            return list(result.scalars().all())
            
        except Exception as e:
            logger.error(f"Error listing overdue tasks: {e}")
            return []
    
    async def get_productivity_stats(
        self,
        db: AsyncSession,
        user_id: str
    ) -> Dict:
        """
        Estatísticas de produtividade (Taskscore).
        
        Args:
            db: Sessão do banco
            user_id: ID do usuário
            
        Returns:
            Dict com total_score, tasks_completed, avg_score
        """
        try:
            # Query para tarefas concluídas
            result = await db.execute(
                select(
                    func.count(DBTask.id).label('tasks_completed'),
                    func.sum(DBTask.score).label('total_score'),
                    func.avg(DBTask.score).label('avg_score')
                )
                .where(
                    DBTask.user_id == user_id,
                    DBTask.status == TaskStatusEnum.CONCLUIDA
                )
            )
            
            row = result.one()
            
            return {
                "tasks_completed": row.tasks_completed or 0,
                "total_score": int(row.total_score or 0),
                "avg_score": float(row.avg_score or 0.0)
            }
            
        except Exception as e:
            logger.error(f"Error getting productivity stats: {e}")
            return {
                "tasks_completed": 0,
                "total_score": 0,
                "avg_score": 0.0
            }


# Instância singleton
task_service = TaskService()
